### ISSUE TITLE
Additional Inputs to Custom Loss Functions

### ISSUE THREAD

**Issue by LudvicLaberge** [dmlc/xgboost#6628]
I've read through the [custom loss tutorial](https://xgboost.readthedocs.io/en/latest/tutorials/custom_metric_obj.html) and was wondering if there is a way to use additional data inputs on top of the labels inside the gradient and hessian functions?

My main objective is being able to provide gradient and hessian functions that are functions of predt, y (=dtrain.get_label()) and **_z_** (some other field in the input data). 

Here is an example of what I had in mind and tried but ran into problems. Here **_z_** represents any input useful to the loss function.
```
# Step 1: Define my own DMatrix class where I can add a z setter/getter

class MyDMatrixWithZ(DMatrix):

    def set_z(self, z): # Inspired from set label, base_margin, etc.
        from xgboost.data import dispatch_meta_backend
        dispatch_meta_backend(self, label, 'label', 'float')

    def get_label(self): # Inspired from get label, base_margin, etc.
        return self.get_float_info('label')

# Step 2: When I create my DMatrix, I use MyDMatrixWithZ instead
dtrain = MyDMatrixWithZ(dataframe[features], label=dataframe[["target"]])
dtrain.set_z(dataframe[["z"]])

# Step 3: Define my custom loss function that requires z
import numpy as np
import xgboost as xgb
from typing import Tuple

def gradient(predt: np.ndarray, dtrain: xgb.DMatrix) -> np.ndarray:
    y = dtrain.get_label()
    z = dtrain.get_z()
    return user_defined_gradient_function(predt, y, z)

def hessian(predt: np.ndarray, dtrain: xgb.DMatrix) -> np.ndarray:
    y = dtrain.get_label()
    z = dtrain.get_z()
    return user_defined_hessian_function(predt, y, z)

def custom_obj(predt: np.ndarray, dtrain: xgb.DMatrix) -> Tuple[np.ndarray, np.ndarray]:
    grad = gradient(predt, dtrain)
    hess = hessian(predt, dtrain)
    return grad, hess

# Step 4: After that follow the exact tutorial posted in Custom Objective Loss Functions and pass custom_obj to obj
```

Hope this is clear, I'd be glad to provide clarifications.

I am a python user and would be glad to contribute changes to the python API and cpp codebase. First I would need to get a sense of if this feature fits within the design orientation of the library to avoid going against the current. But with some guidance I'd be happy to give it a shot.

Thank you

**Comment by trivialfis** [comment]
You can use Python closure to do it:

``` python
some_other_field = ...

def myobj(predt, dtrain):
    y = dtrain.get_label()
    g = user_defined_gradient_function(predt, y, some_other_field)
    h = user_defined_hessian_function(predt, y, some_other_field)
    return g, h
```

**Comment by trivialfis** [comment]
Closing as duplicated to https://github.com/dmlc/xgboost/issues/5990 .

**Comment by hcho3** [comment]
#5990 is closed already? Shouldn't we leave this one open?

**Comment by trivialfis** [comment]
They are similar issues with similar answers.
